### PR TITLE
Add support for multipath device nodes in disk_info

### PR DESCRIPTION
**Pull request by harihare**
added support for multipath device nodes in disk_info tests.

Signed-off-by: Hariharan T S <hari@linux.vnet.ibm.com>

### CODE REVIEW

File: io/disk/disk_info.py [hunk 1]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
+        if "mapper" in self.disk:
+            mpath = True
+            disk_abs = os.path.basename(os.readlink(self.disk))
+            disk = (self.disk.split("/dev/mapper/"))[1]
+            mwwid = multipath.get_multipath_wwid(disk)
+            disk_nodes = multipath.get_paths(mwwid)
+        else:
+            disk_abs = disk = (self.disk.split("/dev/"))[1]
+            disk_nodes.append(disk)
+
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: @harihare Please add this in setup()

File: io/disk/disk_info.py [hunk 2]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
-        if process.system("ls /dev/disk/by-id -l| grep -i %s" % disk,
-                          ignore_status=True, shell=True, sudo=True) != 0:
-            msg.append("Given disk %s is not in /dev/disk/by-id" % disk)
-        if process.system("ls /dev/disk/by-path -l| grep -i %s" % disk,
+        disk_nodes = []
+        mpath = False
+        if "mapper" in self.disk:
+            mpath = True
+            disk_abs = os.path.basename(os.readlink(self.disk))
+            disk = (self.disk.split("/dev/mapper/"))[1]
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: mpath = self.disk.split("/")[-1]
for better naming convention and better way of split

File: io/disk/disk_info.py [hunk 3]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
+            mwwid = multipath.get_multipath_wwid(disk)
+            disk_nodes = multipath.get_paths(mwwid)
+        else:
+            disk_abs = disk = (self.disk.split("/dev/"))[1]
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: disk = self.disk.split("/")[-1]

File: io/disk/disk_info.py [hunk 4]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
+                msg.append("Given disk %s is not in /dev/disk/by-path" % disk_node)
 
         # Verify disk listed in all tools
-        cmd_list = ["fdisk -l ", "parted -l", "lsblk ",
-                    "lshw -c disk "]
+        if mpath:
+            cmd_list = ["fdisk -l ", "lsblk "]
+        else:
+            cmd_list = ["fdisk -l ", "parted -l", "lsblk ",
+                        "lshw -c disk "]
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: operate over disk_node, why mpath again?
we can still use the cmd_list as it is. not require to alter

File: io/disk/disk_info.py [hunk 5]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
             if process.system(cmd, ignore_status=True,
                               shell=True, sudo=True) != 0:
                 msg.append("Given disk %s is not present in %s" % (disk, cmd))
+        if mpath:
+            for disk_node in disk_nodes:
+                if process.system("lshw -c disk | grep -i %s" % disk_node,
+                                  ignore_status=True, shell=True, sudo=True) != 0:
+                    msg.append("Given disk %s is not in lshw -c disk" % disk_node)
+
+
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: same here, operate over disk and not on mpath

File: io/disk/disk_info.py [hunk 6]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
+            disk_abs = os.path.basename(os.readlink(self.disk))
+            disk = (self.disk.split("/dev/mapper/"))[1]
+            mwwid = multipath.get_multipath_wwid(disk)
+            disk_nodes = multipath.get_paths(mwwid)
+        else:
+            disk_abs = disk = (self.disk.split("/dev/"))[1]
+            disk_nodes.append(disk)
+
+        if process.system("ls /dev/disk/by-id -l| grep -i %s" % disk_abs,
                           ignore_status=True, shell=True, sudo=True) != 0:
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: why disk_abs,? only disk is enough

File: io/disk/disk_info.py [hunk 7]
```diff
@@ -217,9 +241,9 @@ def test(self):
             self.log.info("Mount point %s for disk %s updated in df o/p",
                           self.dirs, self.disk)
 
-        if process.system("ls /dev/disk/by-uuid -l| grep -i %s" % disk,
+        if process.system("ls /dev/disk/by-uuid -l| grep -i %s" % disk_abs,
                           ignore_status=True, shell=True, sudo=True) != 0:
-            msg.append("Given disk %s not having uuid" % disk)
+            msg.append("Given disk %s not having uuid" % disk_abs)
```
Comment (Naresh-ibm) [inline io/disk/disk_info.py]: only dis is enough. disk_abs not required.

**Comment by Naresh-ibm** [comment]
@harihare your initiation is right but further implementation requires some modification. lets discuss this tomorrow.

**Comment by harihare** [comment]
[root@ltczz401-lp5 disk]# avocado run disk_info.py -m disk_info.py.data/disk_info.yaml
JOB ID     : 8aaad191eb1957ef16ddd551e588fc92ec8cc212
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2020-02-24T12.51-8aaad19/job.log
 (1/6) disk_info.py:DiskInfo.test_commands;ext4-a7e4: PASS (3.06 s)
 (2/6) disk_info.py:DiskInfo.test_commands;xfs-09a9: PASS (2.86 s)
 (3/6) disk_info.py:DiskInfo.test_commands;btrfs-ed93: CANCEL: btrfs is not supported with                                 RHEL 7.4 onwards (0.17 s)
 (4/6) disk_info.py:DiskInfo.test;ext4-a7e4: PASS (8.17 s)
 (5/6) disk_info.py:DiskInfo.test;xfs-09a9: PASS (5.65 s)
 (6/6) disk_info.py:DiskInfo.test;btrfs-ed93: CANCEL: btrfs is not supported with                                 RHEL 7.4 onwards (0.17 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB TIME   : 20.31 s
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2020-02-24T12.51-8aaad19/results.html

File: io/disk/disk_info.py [hunk 8]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
+            cmd_list = ["fdisk -l ", "parted -l", "lsblk ",
+                        "lshw -c disk "]
         if self.distro == 'Ubuntu':
             cmd_list.append("hwinfo --short --block")
         for cmd in cmd_list:
             cmd = cmd + " | grep -i %s" % disk
             if process.system(cmd, ignore_status=True,
                               shell=True, sudo=True) != 0:
                 msg.append("Given disk %s is not present in %s" % (disk, cmd))
+        if mpath:
```
Comment (narasimhan-v) [inline io/disk/disk_info.py]: Do we need to check this, considering disk_nodes is a list in both cases ?

File: io/disk/disk_info.py [hunk 9]
```diff
@@ -113,24 +115,46 @@ def test(self):
...
         """
         msg = []
-        disk = (self.disk.split("/dev/"))[1]
-        if process.system("ls /dev/disk/by-id -l| grep -i %s" % disk,
-                          ignore_status=True, shell=True, sudo=True) != 0:
-            msg.append("Given disk %s is not in /dev/disk/by-id" % disk)
-        if process.system("ls /dev/disk/by-path -l| grep -i %s" % disk,
+        disk_nodes = []
+        mpath = False
+        if "mapper" in self.disk:
```
Comment (narasimhan-v) [inline io/disk/disk_info.py]: if "dm" in self.disk or "mapper" in self.disk:

**Comment by harihare** [comment]
Resolved all the comments. Please review the same.

**Comment by harihare** [comment]
[root@ltczz401-lp5 disk]# avocado run disk_info.py -m disk_info.py.data/disk_info.yaml
JOB ID     : eddc897cd3075c3c6981b20ea385fcff709d7a13
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2020-03-26T01.08-eddc897/job.log
 (1/6) disk_info.py:DiskInfo.test_commands;ext4-a7e4: PASS (4.57 s)
 (2/6) disk_info.py:DiskInfo.test_commands;xfs-09a9: PASS (4.31 s)
 (3/6) disk_info.py:DiskInfo.test_commands;btrfs-ed93: CANCEL: btrfs is not supported with                                 RHEL 7.4 onwards (0.35 s)
 (4/6) disk_info.py:DiskInfo.test;ext4-a7e4: PASS (8.50 s)
 (5/6) disk_info.py:DiskInfo.test;xfs-09a9: PASS (6.53 s)
 (6/6) disk_info.py:DiskInfo.test;btrfs-ed93: CANCEL: btrfs is not supported with                                 RHEL 7.4 onwards (0.22 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB TIME   : 24.78 s
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2020-03-26T01.08-eddc897/results.html

File: io/disk/disk_info.py [hunk 10]
```diff
@@ -113,24 +126,34 @@ def test(self):
...
         if self.distro == 'Ubuntu':
             cmd_list.append("hwinfo --short --block")
         for cmd in cmd_list:
-            cmd = cmd + " | grep -i %s" % disk
+            cmd = cmd + " | grep -i %s" % self.disk_base
             if process.system(cmd, ignore_status=True,
                               shell=True, sudo=True) != 0:
-                msg.append("Given disk %s is not present in %s" % (disk, cmd))
+                msg.append("Given disk %s is not present in %s" % (self.disk_base, cmd))
+        if self.mpath:
```
[the remark this answers is not in the document]
Comment (narasimhan-v) [reply io/disk/disk_info.py]: @harihare so this check is needed only for multipath, if I understand correctly..

**Comment by harihare** [comment]
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/4385288/job.log)

**Comment by narasimhan-v** [comment]
@Naresh-ibm Thanks for the review.

**Pull request merged**